engine/function/src/templates/Cast.ftl

```diff
@@ -0,0 +1,407 @@
+/*
+ * Copyright (c) 2016-2021 Deephaven Data Labs and Patent Pending
+ */
+
+package io.deephaven.function;
+
+import io.deephaven.vector.*;
+
+import static io.deephaven.util.QueryConstants.*;
+
+/**
+ * Functions for casting between types.
+ */
+@SuppressWarnings("unused")
+public class Cast {
+
+    /**
+     * Cast can not preserve the value.
+     */
+    public static class CastDoesNotPreserveValue extends UnsupportedOperationException {
+
+        private static final long serialVersionUID = 5998276378465685974L;
+
+        /**
+         * Creates a new exception.
+         *
+         * @param message error message.
+         */
+        public CastDoesNotPreserveValue(final String message) {
+            super(message);
+        }
+    }
+
+    private static boolean isLosingPrecisionDouble(final long v) {
+        return v > 9007199254740992L || v < -9007199254740992L;
+    }
+
+    /**
+     * Casts a value to an {@code int}.
+     *
+     * @param value value
+     * @param checkFidelity check to see if the cast preserves the value.
+     * @return cast value.
+     */
+    public static int castInt(byte value, boolean checkFidelity) {
+        return value == NULL_BYTE ? NULL_INT : value;
+    }
+
+    /**
+     * Casts a value to an {@code int}.
+     *
+     * @param value value
+     * @param checkFidelity check to see if the cast preserves the value.
+     * @return cast value.
+     */
+    public static int castInt(short value, boolean checkFidelity) {
+        return value == NULL_SHORT ? NULL_INT : value;
+    }
+
+    /**
+     * Casts a value to an {@code int}.
+     *
+     * @param value value
+     * @param checkFidelity check to see if the cast preserves the value.
+     * @return cast value.
+     */
+    public static int castInt(int value, boolean checkFidelity) {
+        return value;
+    }
+
+    /**
+     * Casts a value to an {@code int}.
+     *
+     * @param value value
+     * @param checkFidelity check to see if the cast preserves the value.
+     * @return cast value.
+     */
+    public static int castInt(long value, boolean checkFidelity) {
+        if( value == NULL_LONG ){
+            return NULL_INT;
+        }
+
```

Comment:
I think you're going to need a second check that looks like this

```
        if (checkFidelity && value == NULL_INT) {
            throw new CastDoesNotPreserveValue("Casting is not supported because the value, " + value + ", represents the null integer sentinel and therefore cannot be cast to int.");
        }
```